BJS2-3417 - Вакансии - Unicorn BC3

hhzktoeto opened the pull request:
(DRAFT) 
Основная логика без тестов

hhzktoeto committed:
Initital commit

hhzktoeto committed:
Реализовал основную логику

Написал методы для создания, обновления и удаления вакансии; для получения вакансий по фильтру и получаения вакансии по ID
Создал интерфейс-фильтр, три фильтра
Создал VacancyMapper и VacancyValidator
Написал один тест на метод create в VacancyService

snailrake commented on src/main/java/faang/school/projectservice/service/vacancy/VacancyService.java:
```diff
@@ -0,0 +1,76 @@
+    private final TeamMemberJpaRepository teamMemberJpaRepository;
+    private final List<VacancyFilter> vacancyFilters;
+
+    public VacancyDto create(VacancyDto vacancyDto) {
+        vacancyValidator.validateIfProjectExistsById(vacancyDto.getProjectId());
+        vacancyValidator.validateCuratorRole(vacancyDto.getCuratorId());
+
+        Vacancy savedVacancy = vacancyRepository.save(vacancyMapper.toEntity(vacancyDto));
+        return vacancyMapper.toDto(savedVacancy);
+    }
```
У вакансии ещё есть статус как обязательное поле. Надо бы дефолтный назначать ей наверное

snailrake commented on src/main/java/faang/school/projectservice/validation/vacancy/VacancyValidator.java:
```diff
@@ -0,0 +1,83 @@
+                .filter(candidate -> candidate.getCandidateStatus().equals(CandidateStatus.ACCEPTED))
+                .count();
+
+        if (candidatesApprovedQuantity <= workersRequiredQuantity) {
```
Почему <=?

hhzktoeto replied:
упс

snailrake commented on src/main/java/faang/school/projectservice/service/vacancy/VacancyService.java:
```diff
@@ -0,0 +1,76 @@
+
+    public List<VacancyDto> getFilteredVacancies(VacancyFilterDto filter) {
+        List<Vacancy> vacancies = vacancyRepository.findAll();
+        if (vacancyFilters != null && !vacancyFilters.isEmpty()) {
```
На null проверка лишняя, даже если не будет фильтров, то будет создан список, просто без бинов

hhzktoeto committed:
Провёл работу по cross-review

hhzktoeto committed:
Решил конфликт

hhzktoeto committed:
Написал тесты на сервис
работает

hhzktoeto committed:
Тесты + мелкие поправки

написал тест на валидатор
исправил в VacancyValidator фильтр на 61 строке, чтобы избежать возможного NullPointerException

IrinaVereshchagina13 commented on src/main/java/faang/school/projectservice/validation/vacancy/VacancyValidator.java:
```diff
@@ -0,0 +1,83 @@
+        teamMembers.forEach(teamMember -> {
+            if (teamMember.getRoles() == null || teamMember.getRoles().isEmpty()) {
+                throw new DataValidationException("Vacancy can't be closed until all team members got their roles");
+            }
```
Это можно включить в первый стрим, чтобы не проходить всех участников еще раз

IrinaVereshchagina13 commented on src/main/java/faang/school/projectservice/validation/vacancy/VacancyValidator.java:
```diff
@@ -0,0 +1,83 @@
+    public void validateIfCandidatesNoMoreNeeded(VacancyDto vacancyDto) {
+        long vacancyId = vacancyDto.getId();
+        int workersRequiredQuantity = vacancyDto.getWorkersRequired();
+        Vacancy vacancy = vacancyRepository.findById(vacancyDto.getId())
+                .orElseThrow(() -> new EntityNotFoundException("Vacancy doesn't exist by id: " + vacancyId));
```
Получение вакансии должно быть в сервисе, здесь только валидация.

IrinaVereshchagina13 commented on src/main/java/faang/school/projectservice/service/vacancy/filter/VacancyNameFilter.java:
```diff
@@ -0,0 +1,18 @@
+
+import java.util.List;
+
+public class VacancyNameFilter implements VacancyFilter {
```
VacancyNameFilter, VacancyDescriptionFilter и VacancyStatusFilter сейчас работать не будут. И они не попадут в List<VacancyFilter> в VacancyService. Есть идеи почему?

hhzktoeto replied:
А бины из них кто делать будет :)

IrinaVereshchagina13 commented on src/main/java/faang/school/projectservice/validation/vacancy/VacancyValidator.java:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    public void validateIfCandidatesNoMoreNeeded(VacancyDto vacancyDto) {
+        long vacancyId = vacancyDto.getId();
```
Это нигде не используется, можно убрать

hhzktoeto committed:
Исправления после review

вынес получение вакансии из валидатора в сервис
объединил всю логику в один стрим в методе validateIfVacancyCanBeClosed в классе VacancyValidator
повесил аннотацию @Component на фильтры
изменил тесты в соответствии с изменениями в коде

hhzktoeto committed:
Мелкие поправки

hhzktoeto committed:
Мелкие поправки
